internal/rbd/snapshot.go

```diff
@@ -0,0 +1,122 @@
+/*
+Copyright 2020 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rbd
+
+import (
+	"context"
+	"strings"
+
+	"github.com/ceph/ceph-csi/internal/util"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+func createRBDClone(ctx context.Context, parentVol, cloneRbdVol *rbdVolume, snap *rbdSnapshot, cr *util.Credentials) error {
+	// create snapshot
+	err := createSnapshot(ctx, parentVol, snap)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to create snapshot %s: %v"), snap, err)
+		return err
+	}
+
+	// create clone image and delete snapshot
+	err = cloneRbdImageFromSnapshot(ctx, cloneRbdVol, snap)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to clone rbd image %s from snapshot %s: %v"), cloneRbdVol.RbdImageName, snap.RbdSnapName, err)
+		err = errors.Errorf("failed to clone rbd image %s from snapshot %s: %v", cloneRbdVol.RbdImageName, snap.RbdSnapName, err)
+	}
+	errSnap := deleteSnapshot(ctx, parentVol, snap)
+	if errSnap != nil {
+		klog.Errorf(util.Log(ctx, "failed to delete snapshot: %v"), errSnap)
+		delErr := deleteImage(ctx, cloneRbdVol, cr)
```

Comment:
If there was an `err` in line 37 above, then deleteImage need not be invoked.

Possibly cleaner to defer `deleteSnapshot` post `createSnapshot` and let the rest of the code play out. Similarly wrap an error routine for defering `deleteImage` (latter is just a thought, former seems more useful)